ttnn/cpp/ttnn/operations/data_movement.hpp

```diff
@@ -0,0 +1,174 @@
+
+    if(dim < rank) {
+        auto the_shape = t.get_shape();
+        auto the_shape_with_padding = t.get_shape().with_tile_padding();
+        return the_shape[dim] != the_shape_with_padding[dim];
+    }
+    return false;
+}
+
+inline ttnn::Tensor permute(
+    const ttnn::Tensor& input_tensor,
+    const std::vector<int>&  order
+) {
+    const bool initial_input_tensor_on_device = is_on_device(input_tensor);
+    const auto input_layout = input_tensor.get_layout();
+    const auto input_rank = input_tensor.get_shape().rank();
+
+    TT_FATAL(input_rank <= 4);
+    TT_FATAL(input_rank == order.size(), "The number of dimensions in the tensor input does not match the length of the desired ordering");
+
+    auto adjust_order = [](const std::vector<int>&  order) {
+        std::vector<std::int64_t>  new_order;
+        TT_FATAL(order.size() <= 4);
+        int additional_ranks = 4 - order.size();
+        for (int i = 0; i < additional_ranks; i++) {
+            new_order.push_back(i);
+        }
+        for (int i = 0; i < order.size(); i++) {
+            new_order.push_back(order.at(i)+additional_ranks);
+        }
+        return new_order;
+    };
+    auto itensor = (input_tensor.get_shape().rank() < 4) ? ttnn::unsqueeze_to_4D(input_tensor) : input_tensor;
+    auto iorder  = adjust_order(order);
+
+    if(has_tile_padding(itensor)) {
+        itensor = ttnn::to_layout(itensor, ttnn::ROW_MAJOR_LAYOUT, std::nullopt, std::nullopt, (Device*)nullptr);
+    }
+
+    TT_FATAL(is_on_device(itensor) and itensor.get_shape().rank() == 4);
+    auto output_tensor = tt::tt_metal::permute(itensor, iorder, ttnn::DRAM_MEMORY_CONFIG);
+    output_tensor = ttnn::to_layout(output_tensor, input_layout, std::nullopt, std::nullopt, (Device*)nullptr);
+
+    if(input_rank < 4){
+        const auto shape = output_tensor.get_shape();
+        const auto full_shape = output_tensor.get_shape().with_tile_padding();
+        std::vector<uint32_t> shape_vec{};
+        std::vector<uint32_t> full_shape_vec{};
+        int i = 0;
+        while(i < 3 and shape[i] == 1) i++;
+        for(i; i < shape.rank(); i++) {
+            shape_vec.push_back(shape[i]);
+            full_shape_vec.push_back(full_shape[i]);
+        }
+        auto metal_shape = tt::tt_metal::Shape(shape_vec, full_shape_vec);
+        output_tensor = ttnn::reshape(output_tensor, ttnn::Shape(metal_shape));
+    }
+
+    if(initial_input_tensor_on_device and not is_on_device(output_tensor)) {
+        output_tensor = ttnn::to_device(output_tensor, input_tensor.device(), ttnn::DRAM_MEMORY_CONFIG);
+    }
+
+    return output_tensor;
+}
+
+inline ttnn::Tensor concat(
+    const std::vector<ttnn::Tensor> & input_tensors,
+    int dim,
+    const std::optional<MemoryConfig>& memory_config_arg
+) {
+    TT_FATAL(input_tensors.size() > 0, "ttnn.concat: expected a non-empty list of Tensors!");
+
+    const auto memory_config = memory_config_arg.value_or(ttnn::DRAM_MEMORY_CONFIG);
```

Comment:
means concat happens on host only?

Reply:
Default output memory config is to create a tensor that lives in DRAM on device. Other options could be on device in L1 interleaved, or on device in L1 sharded.

Been pushing towards standardizing Ops such that this op argument is standardized across all ops.